### PR TITLE
a Bug3964 repairman

### PR DESCRIPTION
**Pull request by maxtaco**
- The strategy is that once we have an unlocked secret key to decrypt
  the LKS client half recovery, to recover a client half, and then
  to see if anything in the local keyring fails to unlock, but can be recovered.
  If so, then to rewrite the local keyring and to rewrite the secret store.
- A test suite to test this fix and the previous recovery fix

### CODE REVIEW

**Comment by maxtaco** [comment]
Note that no code actually calls this engine. We have to figure out where it's best to run it. It requires a login and an unlocked device key.

File: go/engine/bug_3964_repairman.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+// Name provides the name of the engine for the engine interface
+func (b *Bug3964Repairman) Name() string {
+	return "Bug3964Repairman"
+}
+
+// Prereqs returns engine prereqs
+func (b *Bug3964Repairman) Prereqs() Prereqs {
+	return Prereqs{
+		Session: true,
+		Device:  true,
```
Comment (patrickxb) [inline go/engine/bug_3964_repairman.go]: just Device will suffice, but no harm in having both.

File: go/engine/bug_3964_repairman.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+	if dkm, err = b.loadLKSecServerDetails(ctx, lksec); err != nil {
+		return err
+	}
+
+	if ran, err = b.attemptRepair(ctx, lksec, dkm); err != nil {
+		return err
+	}
+	if !ran {
+		return err
```
Comment (patrickxb) [inline go/engine/bug_3964_repairman.go]: `return nil`?

File: go/engine/bug_3964_repairman_test.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+	}, "corruptDevice2")
+	if err != nil {
+		return nil, err
+	}
+	if e2 != nil {
+		return nil, e2
+	}
+	var goodLksec *libkb.LKSec
+
+	// Dev1 has a passphrase cached, but dev2 doesn (since it was provisioned).
```
Comment (patrickxb) [inline go/engine/bug_3964_repairman_test.go]: doesn't

File: go/libkb/login_state.go [hunk 4]
```diff
@@ -1083,6 +1084,25 @@ func (s *LoginState) Keyring(h func(*SKBKeyringFile), name string) error {
...
+func (s *LoginState) MutateKeyring(h func(*SKBKeyringFile) *SKBKeyringFile, name string) error {
+	var err error
+	aerr := s.Account(func(a *Account) {
+		var kr *SKBKeyringFile
+		kr, err = a.Keyring()
+		if err != nil {
+			return
+		}
+		if h(kr) != nil {
+			a.ClearKeyring()
```
Comment (patrickxb) [inline go/libkb/login_state.go]: Could you add a comment that this is just clearing the keyring cache?  Or change function name.  On my first read, I thought it was erasing the keyring.

**Pull request merged**